mate-panel/panel-action-button.c

```diff
@@ -213,13 +217,211 @@ panel_action_logout (GtkWidget *widget)
+			 *with root session managers running
+			 */
+			ret = system ("systemctl poweroff -i");
+			if (ret != 0)
+			{
+				GtkWidget *label;
+				label = gtk_label_new("poweroff not supported from this session");
+				show_error_dialog (label);
+			}
+
+			return;
+		}
+		if (response_id == GTK_RESPONSE_ACCEPT)
+		{
+			gtk_widget_destroy(dialog);
+			system ("reboot now");
+			/*Try the system reboot command first.
+			 *This will fail if root is logged in
+			 *Note that if -f --force was used this would block proper unmounting
+			 *if it does, try systemd in case it's installed as it commonly is
+			 *FIXME: we also need a logind equivalent for non-systemd users
+			 *with root session managers running
+			 */
+			ret = system ("systemctl reboot -i");
+			if (ret != 0)
+			{
+				GtkWidget *label;
+				label = gtk_label_new("reboot not supported from this session");
+				show_error_dialog (label);
+			}
+
+			return;
+		}
+		if (response_id == GTK_RESPONSE_REJECT)
+		{
+			gtk_widget_destroy(dialog);
+			/*FIXME: we also need a logind equivalent for non-systemd users
+			 *with root session managers running
+			 */
+			ret = system ("systemctl hibernate -i");
+			if (ret != 0)
+			{
+				GtkWidget *label;
+				label = gtk_label_new("Hibernation not supported on this system");
+				show_error_dialog (label);
+			}
+
+			return;
+		}
+		if (response_id == GTK_RESPONSE_APPLY)
+		{
+			gtk_widget_destroy(dialog);
+			/*FIXME: we also need a logind equivalent for non-systemd users
+			 *with root session managers running
+			 */
+			ret = system ("systemctl suspend -i");
+			if (ret != 0)
+			{
+				GtkWidget *label;
+				label = gtk_label_new("Suspend not supported on this system");
+				show_error_dialog (label);
+			}
+
+			return;
+		}
+	}
+#endif
+
+static void
+panel_action_shutdown (GtkWidget *widget)
+{
+#ifdef HAVE_WAYLAND
+	GdkDisplay *display = gdk_screen_get_display (gdk_screen_get_default());
+	if (GDK_IS_WAYLAND_DISPLAY (display))
+	{
+		GtkWidget *dialog, *hbox, *image, *label;
+		GtkWidget *suspend_btn, *cancel_btn, *restart_btn, *shutdown_btn, *hibernate_btn;
+		dialog = gtk_dialog_new();
+		int supported, file;
+
+		/*Do we have /sys/power/disk to determine what is supported?*/
+		file = system("cat /sys/power/disk  > /dev/null");
+
+		/*If we do, read it to determine whether to load the button,
+		 *if we do NOT have /sys/power/disk load all buttons unconditionally
+		 */
+		supported = system ("cat /sys/power/disk | grep suspend > /dev/null");
+		if ((supported == 0) || (file !=0))
+		{
+			/*We use the inbuilt gtk response types for simplicity*/
+			suspend_btn = gtk_dialog_add_button (GTK_DIALOG(dialog),
+							     "Suspend",
+							     GTK_RESPONSE_APPLY);
+
+			image = gtk_image_new_from_icon_name("battery", GTK_ICON_SIZE_MENU);
+			gtk_button_set_image (GTK_BUTTON (suspend_btn), image);
+		}
+
+		supported = system ("cat /sys/power/disk | grep hibernate  > /dev/null");
```

Comment:
I think this file controls what the system should do after the hibernation image is writen.
```
[root@mother rave]# cat /sys/power/disk
[platform] shutdown reboot suspend test_resume 
```
see kernel docs https://docs.kernel.org/admin-guide/pm/sleep-states.html#basic-sysfs-interfaces-for-system-suspend-and-hibernation
My notebook w/o hibernation shows me this
```
[rave@satellite ~]$ cat /sys/power/disk 
[disabled]
```
So you should simply grep for suspend?
Or ifnot [disabled]  ?